bin/identify_dormant_github_users.py

```diff
@@ -32,48 +41,150 @@
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+)
 logger = logging.getLogger(__name__)
 
 
-def get_inactive_users_from_data_lake_ignoring_bots_and_collaborators(github_service, bot_list: list) -> list:
-    all_users = github_service.get_all_enterprise_members()
+@dataclass
+class DormantUser:
+    name: str
+    email: str | None
+
+
+def get_usernames_from_csv_ignoring_bots_and_collaborators(bot_list: list) -> list:
+    usernames = []
+    try:
+        with open(CSV_FILE_NAME, mode="r", encoding="utf-8") as file:
+            csv_reader = csv.DictReader(file)
+
+            for row in csv_reader:
+                username = row["login"].strip()
+                is_collaborator = row["outside_collaborator"].lower() == "true"
+                if username not in bot_list and not is_collaborator:
+                    usernames.append(username)
+
+    except FileNotFoundError as e:
+        logging.error("Error reading from file %s: %s", CSV_FILE_NAME, e)
+
+    return usernames
+
+
+def download_github_dormant_users_csv_from_s3():
+    s3 = boto3.client("s3")
+    try:
+        s3.download_file(BUCKET_NAME, CSV_FILE_NAME, CSV_FILE_NAME)
+        logger.info("File %s downloaded successfully.", CSV_FILE_NAME)
+    except NoCredentialsError:
+        logger.error("Credentials not available, please check your AWS credentials.")
+    except FileNotFoundError as e:
+        logger.error("Error downloading file: %s", e)
+
+    if not os.path.isfile(CSV_FILE_NAME):
+        raise FileNotFoundError(
+            f"File {CSV_FILE_NAME} not found in the current directory."
+        )
+
+
+def get_dormant_users_from_github_csv(
+    env_vars: EnvironmentVariables,
+) -> list[DormantUser]:
+    """
+    This function depends on a preliminary manual process: a GitHub Enterprise user
+    must first download the 'dormant.csv' file from the GitHub Enterprise dashboard and
+    then upload it to an S3 bucket. Once this setup is complete, this function will
+    download the 'dormant.csv' file from the S3 bucket and extract a list of
+    usernames from the file.
 
-    cloudtrail_service = CloudtrailService()
-    active_users = cloudtrail_service.get_active_users_for_dormant_users_process()
+    This process ensures that the most current data regarding dormant users is used.
+    """
+    download_github_dormant_users_csv_from_s3()
+    users = get_usernames_from_csv_ignoring_bots_and_collaborators(ALLOWED_BOT_USERS)
 
-    return [user for user in all_users if user not in list(set(active_users).union(bot_list))]
+    # We need to check both the MOJ and AP GitHub organisations as
+    # there is no enterprise opion for this.
+    moj_github_org = GithubService(env_vars.get("GH_MOJ_ADMIN_TOKEN"), MOJ_ORGANISATION)
+    ap_github_org = GithubService(env_vars.get("GH_AP_ADMIN_TOKEN"), AP_ORGANISATION)
```

Comment:
I'd prefer to see these created at the top level and passed through via the parameters - helps to highlight at the top level all dependencies on other services and also which processes require which dependencies 😁 

It also helps the system to fail fast ⏩

Reply:
Cool, this makes sense to me. The readability is sacrificed a little, in my opinion, but you are right that failing early here is the best move.